### PR TITLE
LSQ (Keras QAT)

### PR DESCRIPTION
**Pull request by eladc-git**
Add LSQ algorithm for Keras QAT
LSQ: https://arxiv.org/pdf/1902.08153.pdf

### CODE REVIEW

File: tests/keras_tests/feature_networks_tests/feature_networks/qat/qat_test.py [hunk 1]
```diff
@@ -233,9 +238,7 @@ def run_test(self, **kwargs):
                          input_x=self.representative_data_gen(),
                          quantization_info=quantization_info)
             out_qat_finalize_model = qat_finalize_model(in_tensor)
-            self.unit_test.assertTrue(
-                np.isclose(np.linalg.norm(out_qat_finalize_model - out_ptq_model) / np.linalg.norm(out_ptq_model), 0,
-                           atol=1e-6))
+            self.unit_test.assertTrue(np.isclose(np.linalg.norm(out_qat_finalize_model - out_ptq_model) / np.linalg.norm(out_ptq_model), 0, atol=1e-6))
```
Comment (ofirgo) [inline tests/keras_tests/feature_networks_tests/feature_networks/qat/qat_test.py:241]: same... I think you accidentally reformatted the file, can you revert it?

**Review by ofirgo — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/qat/keras/quantizer/LSQ/__init__.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2023 Sony Semiconductor Israel, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/qat/keras/quantizer/LSQ/__init__.py:1]: Usually, name of algo is in capital letters

File: model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    ActivationPOTInferableQuantizer, ActivationSymmetricInferableQuantizer
+from model_compression_toolkit.trainable_infrastructure.common.base_trainable_quantizer import VariableGroup
+from model_compression_toolkit.qat.keras.quantizer.quant_utils import ste_round, ste_scale
+
+
+def symmetric_quantizer(x, thresholds, num_bits, sign, min_int, max_int, scale_factor):
+    delta = thresholds / (2 ** (num_bits - int(sign)))
+    delta_scaled = ste_scale(delta, scale_factor)
+    rounded = ste_round(x / delta_scaled)
+    clipped = tf.math.minimum(tf.math.maximum(rounded, min_int), max_int)
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py:44]: no

File: tests/keras_tests/feature_networks_tests/feature_networks/qat/qat_test.py [hunk 4]
```diff
@@ -214,8 +220,7 @@ def run_test(self, **kwargs):
                      quantization_info=quantization_info)
 
         out_qat_model = qat_model(in_tensor)
-        self.unit_test.assertTrue(
-            np.isclose(np.linalg.norm(out_qat_model - out_ptq_model) / np.linalg.norm(out_ptq_model), 0, atol=1e-6))
+        self.unit_test.assertTrue(np.isclose(np.linalg.norm(out_qat_model - out_ptq_model) / np.linalg.norm(out_ptq_model), 0, atol=1e-6))
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply tests/keras_tests/feature_networks_tests/feature_networks/qat/qat_test.py:223]: nothing, one line

File: model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    quantized = delta_scaled * clipped
+    return quantized
+
+
+@mark_quantizer(quantization_target=QuantizationTarget.Weights,
+                quantization_method=[QuantizationMethod.POWER_OF_TWO, QuantizationMethod.SYMMETRIC],
+                identifier=TrainingMethod.LSQ)
+class LSQWeightQATQuantizer(BaseKerasQATTrainableQuantizer):
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py:52]: explanation of the quantizer is in TrainingMethod

File: model_compression_toolkit/qat/keras/quantizer/LSQ/uniform_lsq.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                                layer: KerasTrainableQuantizationWrapper):
+        """
+        Add quantizer parameters to the quantizer parameters dictionary
+
+        Args:
+            tensor_shape: tensor shape of the quantized tensor.
+            name: Tensor name.
+            layer: Layer to quantize.
+        """
+        fq_min = layer.add_weight(
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/qat/keras/quantizer/LSQ/uniform_lsq.py:90]: very bad loss

File: model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    """
+    Trainable constrained quantizer to quantize layer's weights.
+    """
+
+    def __init__(self, quantization_config: TrainableQuantizerWeightsConfig):
+        """
+        Initialize a TrainableWeightQuantizer object with parameters to use
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py:59]: I don't think it is necessary

File: model_compression_toolkit/qat/keras/quantizer/LSQ/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/qat/keras/quantizer/LSQ/__init__.py:1]: not in the directory name (and not in other alg names in our project - GPTQ, QAT, STE...). Keep it consistent with the project

[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/qat/keras/quantizer/LSQ/__init__.py:1]: ok, because this is the format in our project

File: model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py]: But what is TrainableWeightQuantizer? it's not even the parent class..

File: model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py]: There is no explanation about LSQ their... just the acronym.
I think it can be helpful if we give a short explanation about the method (or at least a link to the paper)

File: model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+from model_compression_toolkit.qat.keras.quantizer.base_keras_qat_quantizer import BaseKerasQATTrainableQuantizer
+from model_compression_toolkit.trainable_infrastructure import TrainableQuantizerWeightsConfig, \
+    TrainableQuantizerActivationConfig
+from mct_quantizers.keras.quantizers import WeightsPOTInferableQuantizer, WeightsSymmetricInferableQuantizer, \
+    ActivationPOTInferableQuantizer, ActivationSymmetricInferableQuantizer
+from model_compression_toolkit.trainable_infrastructure.common.base_trainable_quantizer import VariableGroup
+from model_compression_toolkit.qat.keras.quantizer.quant_utils import ste_round, ste_scale
+
+
+def symmetric_quantizer(x, thresholds, num_bits, sign, min_int, max_int, scale_factor):
```
Comment (ofirgo) [inline model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py]: Also, typehint, return type and method description

File: model_compression_toolkit/qat/keras/quantizer/quant_utils.py [hunk 9]
```diff
@@ -17,6 +17,23 @@
...
 
+def ste_round(x: tf.Tensor) -> tf.Tensor:
+    """
+    Return the rounded values of a tensor.
+    """
+    error = tf.stop_gradient(tf.math.round(x) - x)
+    return error + x
+
+
+def ste_scale(x: tf.Tensor, scale=1.0) -> tf.Tensor:
```
Comment (elad-c) [inline model_compression_toolkit/qat/keras/quantizer/quant_utils.py:28]: why do we need STE for scaling?

File: model_compression_toolkit/qat/keras/quantizer/lsq/symmetric_lsq.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+    def convert2inferable(self) -> Union[WeightsPOTInferableQuantizer, WeightsSymmetricInferableQuantizer]:
+        """
+        Convert quantizer to inferable quantizer.
+
+        Returns:
+            BaseKerasInferableQuantizer object.
+        """
+        if self.power_of_two:
+            thresholds = 2 ** np.ceil(np.log2(self.get_quantizer_variable(THRESHOLD_TENSOR)))
```
Comment (elad-c) [inline model_compression_toolkit/qat/keras/quantizer/lsq/symmetric_lsq.py:148]: missing **.numpy()** ?

File: model_compression_toolkit/qat/keras/quantizer/lsq/symmetric_lsq.py [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+    def convert2inferable(self) -> Union[ActivationPOTInferableQuantizer, ActivationSymmetricInferableQuantizer]:
+        """
+        Convert quantizer to inferable quantizer.
+
+        Returns:
+            BaseKerasInferableQuantizer object.
+        """
+
+        if self.power_of_two:
+            thresholds = 2 ** np.ceil(np.log2(self.get_quantizer_variable(THRESHOLD_TENSOR)))
```
Comment (elad-c) [inline model_compression_toolkit/qat/keras/quantizer/lsq/symmetric_lsq.py:242]: missing **.numpy()** ?

**Review by elad-c — COMMENTED** [review]
Some small issues

File: model_compression_toolkit/qat/common/qat_config.py [hunk 12]
```diff
@@ -45,9 +45,12 @@ class TrainingMethod(Enum):
 
     DQA -  DNN Quantization with Attention. Includes a smooth quantization introduces by DQA method
 
+    LSQ - Learned Step size Quantization. Includes PowerOfTwo, symmetric & uniform quantizers
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/qat/common/qat_config.py]: I will add

File: model_compression_toolkit/qat/keras/quantizer/quant_utils.py (hunk 9 above)
Comment (eladc-git) [reply model_compression_toolkit/qat/keras/quantizer/quant_utils.py:28]: This is one of the component of LSQ: STE for the scale so it will influence only on backward pass)

File: model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py]: so it should be LSQWeightQATQuantizer (not LSQ)

[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/qat/keras/quantizer/LSQ/symmetric_lsq.py]: Yes that's what I meant, sorry if that wasn't clear

**Review by elad-c — APPROVED** [review]

**Review by ofirgo — APPROVED** [review]
